File: programmers/js/new/level2/쿼드압축_후_개수_세기.js
function solution(arr) {
    const check = Array.from(Array(arr.length), () => Array(arr.length).fill(false));
    dfs(arr, 0, arr.length, 0, arr.length, check);
    const count = [0, 0];
    check.forEach((row, i) => {
        row.forEach((c, j) => {
            if (c) {
                count[arr[i][j]]++;
            }
        });
    });
    return count;
}

function dfs(arr, t, b, l, r, check) {
    if (isAllEqual(arr, t, b, l, r)) {
        checkRemainFalse(check, t, b, l, r);
        return;
    }
    const rm = parseInt((t + b) / 2);
    const cm = parseInt((l + r) / 2);
    // 1
    if (isAllEqual(arr, t, rm, l, cm)) {
        checkRemainFalse(check, t, rm, l, cm);
    } else {
        dfs(arr, t, rm, l, cm, check);
    }
    // 2
    if (isAllEqual(arr, t, rm, cm, r)) {
        checkRemainFalse(check, t, rm, cm, r);
    } else {
        dfs(arr, t, rm, cm, r, check);
    }
    // 3
    if (isAllEqual(arr, rm, b, l, cm)) {
        checkRemainFalse(check, rm, b, l, cm);
    } else {
        dfs(arr, rm, b, l, cm, check);
    }
    // 4
    if (isAllEqual(arr, rm, b, cm, r)) {
        checkRemainFalse(check, rm, b, cm, r);
    } else {
        dfs(arr, rm, b, cm, r, check);
    }
}

function isAllEqual(arr, t, b, l, r) {
    const val = arr[t][l];
    for (let i = t; i < b; ++i) {
        for (let j = l; j < r; ++j) {
            if (arr[i][j] !== val) {
                return false;
            }
        }
    }
    return true;
}

function checkRemainFalse(check, t, b, l, r) {
    for (let i = t; i < b; ++i) {
        for (let j = l; j < r; ++j) {
            check[i][j] = false;
        }
    }
    check[t][l] = true;
}

console.log(
    solution([
        [1, 1, 0, 0],
        [1, 0, 0, 0],
        [1, 0, 0, 1],
        [1, 1, 1, 1],
    ])
);

File: programmers/python/level2/쿼드압축_후_개수_세기.py
def solution(arr):
    check_list = [[True for j in range(len(arr))] for i in range(len(arr))]
    compress(arr, 0, 0, len(arr), check_list)
    count = [0,0]
    for i in range(len(check_list)):
        for j in range(len(check_list)):
            if check_list[i][j]:
                count[arr[i][j]] += 1
    return count

def compress(arr, r,c,l, check_list):
    if is_all_equal(arr, r, c, l):
        check_false(check_list, r, c, l)
    else:
        m = l // 2
        # 1사분면
        compress(arr, r, c+m, m, check_list)
        # 2사분면
        compress(arr, r,c,m, check_list)
        # 3사분면
        compress(arr, r+m, c, m, check_list)
        # 4사분면
        compress(arr, r+m, c+m, m, check_list)

def check_false(check_list, r, c, l):
    for i in range(r, r+l):
        for j in range(c, c+l):
            check_list[i][j] = False
    check_list[i][j] = True

def is_all_equal(arr, r,c,l):
    val = arr[r][c]
    for i in range(r, r+l):
        for j in range(c, c+l):
            if val != arr[i][j]:
                return False
    return True

print(solution([[1,1,0,0],[1,0,0,0],[1,0,0,1],[1,1,1,1]]))